src/resolve/resolved-hook.c

```diff
@@ -0,0 +1,852 @@
+
+        /* timestamp we last saw this in CLOCK_MONOTONIC, for GC handling */
+        uint64_t seen_usec;
+
+        /* When a hook never responds correctly, we'll eventually give up trying */
+        RateLimit reconnect_ratelimit;
+} Hook;
+
+static Hook* hook_free(Hook *h) {
+        if (!h)
+                return NULL;
+
+        mfree(h->socket_path);
+        sd_varlink_unref(h->filter_link);
+        set_free(h->idle_links);
+
+        set_free(h->filter_domains);
+
+        return mfree(h);
+}
+
+DEFINE_PRIVATE_TRIVIAL_REF_UNREF_FUNC(Hook, hook, hook_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(Hook*, hook_unref);
+
+static Hook *hook_unlink(Hook *h) {
+        if (!h)
+                return NULL;
+
+        if (!h->manager)
+                return NULL;
+
+        if (h->socket_path)
+                hashmap_remove(h->manager->hooks, h->socket_path);
+        h->manager = NULL;
+
+        return hook_unref(h);
+}
+
+static int dispatch_filter_domains(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        Hook *h = ASSERT_PTR(userdata);
+        int r;
+
+        if (!sd_json_variant_is_array(variant))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not an array.", strna(name));
+
+        /* Let's explicitly allocate the set here, since we want that a NULL set means: let everything
+         * through; but an empty set shall mean: let nothing through */
+        r = set_ensure_allocated(&h->filter_domains, &dns_name_hash_ops_free);
+        if (r < 0)
+                return json_log_oom(variant, flags);
+
+        sd_json_variant *i;
+        JSON_VARIANT_ARRAY_FOREACH(i, variant) {
+                if (!sd_json_variant_is_string(i))
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "Element of JSON field '%s' is not a string.", strna(name));
+
+                r = set_put_strdup_full(&h->filter_domains, &dns_name_hash_ops_free, sd_json_variant_string(i));
+                if (r < 0 && r != -EEXIST)
+                        return json_log_oom(variant, flags);
+        }
+
+        return 0;
+}
+
+static void hook_reset_filter(Hook *h) {
+        assert(h);
+
+        h->filter_domains = set_free(h->filter_domains);
+        h->filter_labels_min = UINT_MAX;
+        h->filter_labels_max = UINT_MAX;
+}
+
+static int hook_acquire_filter(Hook *h);
+
+static int on_filter_reply(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        Hook *h = ASSERT_PTR(userdata);
+        int r;
+
+        if (error_id) {
+                if (streq(error_id, SD_VARLINK_ERROR_DISCONNECTED)) {
+                        if (ratelimit_below(&h->reconnect_ratelimit)) {
```

Comment:
Again here, I'm not sure what the purpose of the ratelimit is. If the hook fails to reply to the initial query, then almost certainly it is nonfunctional and we don't want to have anything to do with it, no?